pkg/dashboards/folders.go

```diff
@@ -0,0 +1,883 @@
+	breadcrumbs := make([]Breadcrumb, 0)
+	currentID := folderID
+
+	for currentID != "" {
+		item, exists := structure.Items[currentID]
+		if !exists {
+			break
+		}
+
+		breadcrumbs = append([]Breadcrumb{{
+			ID:   currentID,
+			Name: item.Name,
+		}}, breadcrumbs...)
+
+		currentID = item.ParentID
+	}
+
+	return breadcrumbs
+}
+
+func updateFolder(folderID string, req *UpdateFolderRequest) error {
+	if folderID == rootFolderID {
+		return fmt.Errorf("cannot update root folder")
+	}
+
+	if isDefaultFolder(folderID) {
+		return fmt.Errorf("cannot update default folder")
+	}
+
+	structure, err := readFolderStructure()
+	if err != nil {
+		return fmt.Errorf("failed to read folder structure: %v", err)
+	}
+
+	folder, exists := structure.Items[folderID]
+	if !exists {
+		return fmt.Errorf("folder not found: %s", folderID)
+	}
+
+	// If moving to new parent
+	if req.ParentID != "" && req.ParentID != folder.ParentID {
+		// Validate new parent exists and is a folder
+		newParent, exists := structure.Items[req.ParentID]
+		if !exists {
+			return fmt.Errorf("new parent folder not found: %s", req.ParentID)
+		}
+		if newParent.Type != "folder" {
+			return fmt.Errorf("new parent must be a folder")
+		}
+
+		// Check for circular reference
+		if wouldCreateCircularReference(folderID, req.ParentID, structure) {
+			return fmt.Errorf("cannot move folder: would create circular reference")
+		}
+
+		// Remove from old parent's order
+		if oldParentOrder, exists := structure.Order[folder.ParentID]; exists {
+			newOrder := make([]string, 0)
+			for _, id := range oldParentOrder {
+				if id != folderID {
+					newOrder = append(newOrder, id)
+				}
+			}
+			structure.Order[folder.ParentID] = newOrder
+		}
+
+		// Add to new parent's order
+		structure.Order[req.ParentID] = append(structure.Order[req.ParentID], folderID)
+
+		// Update parent ID
+		folder.ParentID = req.ParentID
+	}
+
+	// Update name if provided
+	if req.Name != "" && req.Name != folder.Name {
+		// Check for duplicate names in the same parent
+		parentID := folder.ParentID
+		if req.ParentID != "" {
+			parentID = req.ParentID
+		}
+
+		for _, siblingID := range structure.Order[parentID] {
+			if sibling, exists := structure.Items[siblingID]; exists {
+				if sibling.Name == req.Name && siblingID != folderID {
+					return fmt.Errorf("folder with name %s already exists in this location", req.Name)
+				}
+			}
+		}
+		folder.Name = req.Name
+	}
+
+	structure.Items[folderID] = folder
+
+	if err := writeFolderStructure(structure); err != nil {
+		return fmt.Errorf("failed to write folder structure: %v", err)
+	}
+
+	return nil
+}
+
+func wouldCreateCircularReference(folderID, newParentID string, structure *FolderStructure) bool {
+	current := newParentID
+	for current != "" {
+		if current == folderID {
+			return true
+		}
+		if item, exists := structure.Items[current]; exists {
+			current = item.ParentID
+		} else {
+			break
```

Comment:
I think this will cause an infinite loop if there's already a circular reference. That might be ok, since I think we'll never get into a state where there's already a circular reference when we call this function though, right? If we do need to be more robust and handle this case, you can lookup the classic solution to checking if a linked list has a cycle